File: ios/Runner/RtcFramesPlugin.swift
import Flutter
import UIKit
import WebRTC

/// Native companion for the RTC v2 (P2P WebRTC) path.
///
/// The app owns the camera (CameraFrameService feeds the AI scorer), so the
/// WebRTC video track cannot use a WebRTC-managed capturer. This plugin
/// reaches into flutter_webrtc's singleton to create an RTCVideoSource-backed
/// track with a fixed id, registers it in the plugin's `localTracks` (exactly
/// like getUserMedia registers its camera track, LocalVideoTrack wrapper
/// included) so every Dart-side call (addTrack / replaceTrack / trackDispose)
/// resolves it, and exposes a `pushFrame` channel the camera service feeds
/// with the same BGRA buffers it pushes to Agora.
///
/// Registered from AppDelegate on the `com.dartrivals/rtc_frames` channel —
/// same pattern as NativeInferencePlugin.
class RtcFramesPlugin: NSObject {
  static let trackId = "dartrivals-ext-cam"

  private var videoSource: RTCVideoSource?
  private var videoTrack: RTCVideoTrack?
  private var capturer: RTCVideoCapturer?
  // Reused staging buffer. Only ever touched on frameQueue: the queue is
  // serial, so the buffer is always fully consumed (toI420 copy) before the
  // next frame reuses it — a single one is safe.
  private var pixelBuffer: CVPixelBuffer?

  // The memcpy + libyuv toI420 of a ~1.5MB BGRA frame at ≤16fps is real CPU;
  // doing it on the platform thread competed with UI + every other channel
  // (and this codebase has an overheating history). Serial: preserves frame
  // order and the pixel-buffer reuse invariant. The FlutterResult is answered
  // only after the frame is consumed, so the Dart-side 2-in-flight cap keeps
  // acting as backpressure (excess frames drop client-side, nothing queues).
  private let frameQueue = DispatchQueue(label: "com.dartrivals.rtcframes", qos: .userInitiated)

  func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "create":
      create(result: result)
    case "pushFrame":
      pushFrame(call, result: result)
    case "dispose":
      disposeTrack(result: result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func create(result: @escaping FlutterResult) {
    guard let plugin = FlutterWebRTCPlugin.sharedSingleton() else {
      result(FlutterError(code: "no_plugin",
                          message: "flutter_webrtc plugin not registered",
                          details: nil))
      return
    }
    // The factory is created by the plugin's own 'initialize' method call;
    // the Dart side (RtcFramesService.createTrack) forces WebRTC.initialize()
    // before invoking us, so this should always be non-nil.
    guard let factory = plugin.peerConnectionFactory else {
      result(FlutterError(code: "no_factory",
                          message: "peerConnectionFactory not initialized",
                          details: nil))
      return
    }
    if videoTrack == nil {
      let source = factory.videoSource()
      let track = factory.videoTrack(with: source, trackId: RtcFramesPlugin.trackId)
      track.isEnabled = true
      videoSource = source
      videoTrack = track
      // RTCVideoSource is itself the RTCVideoCapturerDelegate; the capturer
      // instance only identifies the frame producer in the delegate call.
      capturer = RTCVideoCapturer(delegate: source)
    }
    plugin.localTracks?[RtcFramesPlugin.trackId] = LocalVideoTrack(track: videoTrack!)
    result(RtcFramesPlugin.trackId)
  }

  /// args: bytes (BGRA), width, height, rotation (degrees), bytesPerRow
  /// (source row stride — camera buffers can be row-padded).
  private func pushFrame(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let source = videoSource,
          let capturer = capturer,
          let args = call.arguments as? [String: Any],
          let bytes = args["bytes"] as? FlutterStandardTypedData,
          let width = args["width"] as? Int,
          let height = args["height"] as? Int,
          width > 0, height > 0
    else {
      result(nil)
      return
    }
    let rotation = (args["rotation"] as? Int) ?? 0
    let srcBytesPerRow = (args["bytesPerRow"] as? Int) ?? (width * 4)
    // `bytes.data` (immutable NSData) and source/capturer are captured by the
    // block, so a concurrent disposeTrack can't pull them out from under us.
    let data = bytes.data

    frameQueue.async { [weak self] in
      defer { DispatchQueue.main.async { result(nil) } }
      guard let self = self else { return }

      var pb: CVPixelBuffer?
      if let existing = self.pixelBuffer,
         CVPixelBufferGetWidth(existing) == width,
         CVPixelBufferGetHeight(existing) == height {
        pb = existing
      } else {
        CVPixelBufferCreate(kCFAllocatorDefault, width, height,
                            kCVPixelFormatType_32BGRA, nil, &pb)
        self.pixelBuffer = pb
      }
      guard let pixelBuffer = pb else { return }

      CVPixelBufferLockBaseAddress(pixelBuffer, [])
      if let dst = CVPixelBufferGetBaseAddress(pixelBuffer) {
        let dstBytesPerRow = CVPixelBufferGetBytesPerRow(pixelBuffer)
        let rowBytes = min(width * 4, min(srcBytesPerRow, dstBytesPerRow))
        data.withUnsafeBytes { (src: UnsafeRawBufferPointer) in
          guard let srcBase = src.baseAddress else { return }
          for row in 0..<height {
            let srcOff = row * srcBytesPerRow
            guard srcOff + rowBytes <= src.count else { break }
            memcpy(dst + row * dstBytesPerRow, srcBase + srcOff, rowBytes)
          }
        }
      }
      CVPixelBufferUnlockBaseAddress(pixelBuffer, [])

      // toI420 (libyuv) detaches the frame from our reused pixel buffer
      // before it crosses into the encoder, and hands the encoder its
      // preferred format regardless of hardware/software H264 path.
      let rtcBuffer = RTCCVPixelBuffer(pixelBuffer: pixelBuffer)
      let rtcRotation = RTCVideoRotation(rawValue: rotation) ?? ._0
      let timeNs = Int64(CACurrentMediaTime() * 1_000_000_000)
      let frame = RTCVideoFrame(buffer: rtcBuffer.toI420(),
                                rotation: rtcRotation,
                                timeStampNs: timeNs)
      source.capturer(capturer, didCapture: frame)
    }
  }

  private func disposeTrack(result: @escaping FlutterResult) {
    if let plugin = FlutterWebRTCPlugin.sharedSingleton() {
      plugin.localTracks?.removeObject(forKey: RtcFramesPlugin.trackId)
    }
    // Nil the ivars ON THE PLATFORM THREAD: a stale teardown block racing on
    // frameQueue must never undo a create() that ran meanwhile (the
    // deferred-dispose overlap), and main-thread reads of these ivars must
    // not race frameQueue writes. In-flight frame blocks captured
    // source/capturer locally and finish safely (ARC keeps them alive).
    videoTrack = nil
    videoSource = nil
    capturer = nil
    // Only pixelBuffer belongs to the frameQueue domain.
    frameQueue.async { [weak self] in
      self?.pixelBuffer = nil
      DispatchQueue.main.async { result(nil) }
    }
  }
}

File: android/app/src/main/kotlin/com/dartrivals/app/RtcFramesPlugin.kt
package com.dartrivals.app

import android.os.Handler
import android.os.HandlerThread
import android.os.Looper
import android.os.SystemClock
import com.cloudwebrtc.webrtc.FlutterWebRTCPlugin
import com.cloudwebrtc.webrtc.MethodCallHandlerImpl
import com.cloudwebrtc.webrtc.video.LocalVideoTrack
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import org.webrtc.JavaI420Buffer
import org.webrtc.NV21Buffer
import org.webrtc.VideoFrame
import org.webrtc.VideoSource
import org.webrtc.VideoTrack
import java.nio.ByteBuffer
import kotlin.math.min

/**
 * Native companion for the RTC v2 (P2P WebRTC) path.
 *
 * The app owns the camera (CameraFrameService feeds the AI scorer), so the
 * WebRTC video track cannot use a WebRTC-managed capturer. This plugin reaches
 * into flutter_webrtc's singleton to create a VideoSource-backed track with a
 * fixed id, registers it in the plugin's localTracks (same LocalVideoTrack
 * wrapper getUserMedia uses) so every Dart-side call (addTrack / replaceTrack /
 * trackDispose) resolves it, and exposes a `pushFrame` channel the camera
 * service feeds with the same tight I420/NV21 buffers it pushes to Agora.
 *
 * Registered from MainActivity — same pattern as NativeInferencePlugin.
 * org.webrtc classes come from `compileOnly io.github.webrtc-sdk:android`
 * (see app build.gradle.kts) — at runtime they are provided by flutter_webrtc's
 * own dependency, so the versions must stay in lock-step.
 */
class RtcFramesPlugin(
    // The FlutterWebRTCPlugin instance attached to OUR engine, resolved by
    // MainActivity from the engine's plugin registry. NEVER use
    // FlutterWebRTCPlugin.sharedSingleton here: its constructor overwrites
    // the static on every instantiation, and the push-notification background
    // isolate registers a second instance whose factory is never initialized
    // — reaching through the singleton returned no_factory while the main
    // engine's own instance was perfectly healthy (2026-08-02 prod test).
    private val webrtcPlugin: FlutterWebRTCPlugin?,
) : FlutterPlugin, MethodChannel.MethodCallHandler {

    companion object {
        const val TRACK_ID = "dartrivals-ext-cam"
    }

    private lateinit var channel: MethodChannel
    private var videoSource: VideoSource? = null
    private var videoTrack: VideoTrack? = null

    // NV21→I420 (libyuv full-frame conversion + a 1.4MB direct-buffer
    // allocation) runs on the CALLING thread inside libwebrtc's
    // onFrameCaptured — verified in the 144.7559.09 AAR bytecode. At 15fps
    // on the platform thread that competed with UI + every other channel
    // (and this codebase has an overheating history). Serial HandlerThread =
    // frame order preserved; the result is answered only after the frame is
    // consumed, so the Dart-side 2-in-flight cap keeps acting as
    // backpressure. Mirrors the iOS frameQueue.
    private val frameThread = HandlerThread("dartrivals-rtc-frames").apply { start() }
    private val frameHandler = Handler(frameThread.looper)
    private val mainHandler = Handler(Looper.getMainLooper())

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, "com.dartrivals/rtc_frames")
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        release()
        // After release()'s posted disposals drain.
        frameThread.quitSafely()
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "create" -> create(result)
            "pushFrame" -> pushFrame(call, result)
            "dispose" -> {
                disposeTrack()
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    /**
     * flutter_webrtc keeps its MethodCallHandlerImpl (the StateProvider owning
     * localTracks and the PeerConnectionFactory) in a private field with no
     * public accessor for putLocalTrack, so this is the one reflective seam.
     * Safe under R8: the plugin ships `-keep class com.cloudwebrtc.webrtc.**`
     * consumer rules, so the field name survives minification.
     */
    private fun webrtcHandler(): MethodCallHandlerImpl? {
        // No sharedSingleton fallback: that static is exactly the
        // background-isolate-hijacked instance documented above — better to
        // fail loudly into telemetry than silently reach the wrong engine.
        val plugin = webrtcPlugin ?: return null
        return try {
            val field = FlutterWebRTCPlugin::class.java.getDeclaredField("methodCallHandler")
            field.isAccessible = true
            field.get(plugin) as? MethodCallHandlerImpl
        } catch (e: Exception) {
            null
        }
    }

    private fun create(result: MethodChannel.Result) {
        val handler = webrtcHandler()
        if (handler == null) {
            result.error("no_handler", "flutter_webrtc plugin/handler unreachable", null)
            return
        }
        // The factory is created by flutter_webrtc's native 'initialize',
        // which has run by now because the Dart side creates the
        // RTCPeerConnection before asking for this track.
        val factory = handler.peerConnectionFactory
        if (factory == null) {
            result.error("no_factory", "flutter_webrtc factory not initialized", null)
            return
        }
        if (videoTrack == null) {
            val source = factory.createVideoSource(false)
            // Canonical custom-capture contract: mark the source live before
            // frames arrive — a source never "started" can drop every frame.
            source.capturerObserver.onCapturerStarted(true)
            val track = factory.createVideoTrack(TRACK_ID, source)
            track.setEnabled(true)
            videoSource = source
            videoTrack = track
        }
        handler.putLocalTrack(TRACK_ID, LocalVideoTrack(videoTrack!!))
        result.success(TRACK_ID)
    }

    /**
     * args: bytes (tight planar I420 or NV21, no row padding), width, height,
     * rotation (clockwise degrees to upright — same semantic as the Agora
     * push), format ('i420' | 'nv21').
     */
    private fun pushFrame(call: MethodCall, result: MethodChannel.Result) {
        val source = videoSource
        val bytes = call.argument<ByteArray>("bytes")
        val width = call.argument<Int>("width")
        val height = call.argument<Int>("height")
        if (source == null || bytes == null || width == null || height == null ||
            width <= 0 || height <= 0
        ) {
            result.success(null)
            return
        }
        val rotation = call.argument<Int>("rotation") ?: 0
        val format = call.argument<String>("format") ?: "nv21"
        // libyuv reads w*h*1.5 bytes unconditionally on the NV21 path — a
        // short buffer would be a NATIVE overread (SIGSEGV, not a Java
        // exception the catch below could see).
        if (bytes.size < width * height * 3 / 2) {
            result.success(null)
            return
        }

        frameHandler.post {
            try {
                val buffer: VideoFrame.Buffer = if (format == "i420") {
                    // JavaI420Buffer.allocate uses tight strides (y=width,
                    // u/v=width/2 for even dimensions), matching our packed
                    // input.
                    val i420 = JavaI420Buffer.allocate(width, height)
                    val ySize = width * height
                    val uvSize = (width / 2) * (height / 2)
                    putPlane(bytes, 0, ySize, i420.dataY)
                    putPlane(bytes, ySize, uvSize, i420.dataU)
                    putPlane(bytes, ySize + uvSize, uvSize, i420.dataV)
                    i420
                } else {
                    // Zero-copy wrap: the byte[] decoded from the method
                    // channel is exclusively ours (conversion happens in
                    // onFrameCaptured, on THIS thread).
                    NV21Buffer(bytes, width, height, null)
                }
                val frame = VideoFrame(buffer, rotation, SystemClock.elapsedRealtimeNanos())
                source.capturerObserver.onFrameCaptured(frame)
                frame.release()
            } catch (_: Exception) {
                // A malformed frame must never take the channel down; the
                // next frame is at most ~60ms away.
            }
            // FlutterJNI requires responses on the platform thread; answering
            // only after consumption keeps the Dart 2-in-flight backpressure.
            mainHandler.post { result.success(null) }
        }
    }

    private fun putPlane(src: ByteArray, offset: Int, length: Int, dst: ByteBuffer) {
        val d = dst.duplicate()
        d.clear()
        d.put(src, offset, min(length, min(src.size - offset, d.remaining())))
    }

    private fun disposeTrack() {
        // Removes the entry from localTracks (does NOT dispose the track).
        try {
            webrtcHandler()?.trackDispose(TRACK_ID)
        } catch (_: Exception) {
        }
        release()
    }

    private fun release() {
        // Nil the fields ON THE PLATFORM THREAD (a subsequent create() can't
        // be undone by a stale teardown), but run the actual disposals on
        // frameHandler so any in-flight frame finishes against a live
        // source first (serial queue = ordered after it).
        val track = videoTrack
        val source = videoSource
        videoTrack = null
        videoSource = null
        frameHandler.post {
            try {
                track?.dispose()
            } catch (_: Exception) {
            }
            try {
                source?.dispose()
            } catch (_: Exception) {
            }
        }
    }
}
